### PR TITLE
Draft function to take full MC step

### PR DESCRIPTION
**Pull request by ffoucart**
## Proposed changes

Adds function responsible for taking a full MC time step during evolution

### Upgrade instructions

None

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

**Comment by ffoucart** [comment]
Currently only here as a guide to see how the various functions and variables are supposed to interact, and to see what remains to be coded. This PR depends on #5902  and #5904, and will conflict with #5851, #5870, and #5903.

**Comment by ffoucart** [comment]
Rebased against 5902 and ready for review.

File: src/Evolution/Particles/MonteCarlo/TemplatedLocalFunctions.hpp [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 #include "DataStructures/DataVector.hpp"
 #include "DataStructures/Tensor/Tensor.hpp"
+#include "Evolution/Particles/MonteCarlo/NeutrinoInteractionTable.hpp"
```
Comment (kidder) [inline src/Evolution/Particles/MonteCarlo/TemplatedLocalFunctions.hpp]: Anything with a `std::` in this file needs the appropriate include.  All of the spectre includes should just be forward declarations and an include of `DataStructures/Tensor/TypeAliases.hpp`.  You may need to add these includes to the cpp file.

File: tests/Unit/Evolution/Particles/MonteCarlo/Test_TakeTimeStep.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+  MonteCarloStruct.take_time_step_on_element(
+      &packets, &generator, &single_packet_energy, start_time, final_time,
+      interaction_table, electron_fraction, baryon_density, temperature,
+      lorentz_factor, lower_spatial_four_velocity, lapse, shift, d_lapse,
+      d_shift, d_inv_spatial_metric, spatial_metric, inv_spatial_metric,
+      determinant_spatial_metric, mesh, mesh_coordinates, mesh_velocity,
+      inverse_jacobian_logical_to_inertial, det_jacobian_logical_to_inertial,
+      jacobian_inertial_to_fluid, inverse_jacobian_inertial_to_fluid);
+}
```
Comment (kidder) [inline tests/Unit/Evolution/Particles/MonteCarlo/Test_TakeTimeStep.cpp:158]: is there any property of the result that you can CHECK the value of?

File: src/Evolution/Particles/MonteCarlo/TakeTimeStep.tpp [hunk 3]
```diff
@@ -0,0 +1,125 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include "Evolution/Particles/MonteCarlo/CellVolume.hpp"
+#include "Evolution/Particles/MonteCarlo/TemplatedLocalFunctions.hpp"
```
Comment (kidder) [inline src/Evolution/Particles/MonteCarlo/TakeTimeStep.tpp:6]: make this the first include, with a blank line after it. 

Also add the appropriate includes

File: src/Evolution/Particles/MonteCarlo/TakeTimeStep.tpp [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+
+template <size_t EnergyBins, size_t NeutrinoSpecies>
+void TemplatedLocalFunctions<EnergyBins, NeutrinoSpecies>::
+    take_time_step_on_element(
+        const gsl::not_null<std::vector<Packet>*> packets,
+        const gsl::not_null<std::mt19937*> random_number_generator,
+        const gsl::not_null<std::array<DataVector, NeutrinoSpecies>*>
+            single_packet_energy,
+
+        const double& start_time, const double& target_end_time,
```
Comment (kidder) [inline src/Evolution/Particles/MonteCarlo/TakeTimeStep.tpp]: const double& -> const double

File: src/Evolution/Particles/MonteCarlo/TemplatedLocalFunctions.hpp [hunk 5]
```diff
@@ -24,18 +25,56 @@ struct Packet;
...
+   * \brief Function to take a single Monte Carlo time step on a
+   * finite difference element.
+   */
+  void take_time_step_on_element(
+      gsl::not_null<std::vector<Packet>*> packets,
+      gsl::not_null<std::mt19937*> random_number_generator,
+      gsl::not_null<std::array<DataVector, NeutrinoSpecies>*>
+          single_packet_energy,
+
+      const double& start_time, const double& target_end_time,
```
Comment (kidder) [inline src/Evolution/Particles/MonteCarlo/TemplatedLocalFunctions.hpp]: const double& -> double

**Review by kidder — CHANGES_REQUESTED** [review]
you can squash the changes

**Comment by ffoucart** [comment]
Rebased, addressed code review comments, and squashed.

**Review by kidder — APPROVED** [review]

**Pull request merged**